datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java

```diff
@@ -0,0 +1,153 @@
+ * assignment. This StickyPartitionAssignmentStrategy creates new tasks and remove old tasks to accommodate the
+ * change in partition assignment. The strategy is also "Sticky", i.e., it minimizes the potential task mutations.
+ * The total number of tasks is also unchanged during this process.
+ */
+public class StickyPartitionAssignmentStrategy extends StickyMulticastStrategy {
+  private static final Logger LOG = LoggerFactory.getLogger(StickyPartitionAssignmentStrategy.class.getName());
+
+  /**
+   * Constructor for StickyPartitionAssignmentStrategy
+   * @param maxTasks Maximum number of {@link DatastreamTask}s to create out
+   *                 of any {@link com.linkedin.datastream.common.Datastream}
+   *                 if no value is specified for the "maxTasks" config property
+   *                 at an individual datastream level.
+   * @param imbalanceThreshold The maximum allowable difference in the number of tasks assigned
+   *                           between any two {@link com.linkedin.datastream.server.Coordinator}
+   *                           instances, before triggering a rebalance. The default is
+   *                           {@value DEFAULT_IMBALANCE_THRESHOLD}.
+   */
+  public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<Integer> imbalanceThreshold) {
+    super(maxTasks, imbalanceThreshold);
+  }
+  /**
+   * assign partitions to a particular datastream group
+   *
+   * @param currentAssignment the old assignment
+   * @param datastreamPartitions the subscribed partitions for the particular datastream group
+   * @return new assignment mapping
+   */
+  public Map<String, Set<DatastreamTask>> assignPartitions(Map<String,
+      Set<DatastreamTask>> currentAssignment, DatastreamGroupPartitionsMetadata datastreamPartitions) {
+
+    LOG.debug("old partition assignment info, assignment: {}", currentAssignment);
+
+    String dgName = datastreamPartitions.getDatastreamGroup().getName();
+
+    // Step 1: collect the # of tasks and figured out the unassigned partitions
+    List<String> assignedPartitions = new ArrayList<>();
+    int totalTaskCount = 0;
+    for (Set<DatastreamTask> tasks : currentAssignment.values()) {
+      Set<DatastreamTask> dgTask = tasks.stream().filter(t -> dgName.equals(t.getTaskPrefix())).collect(Collectors.toSet());
+      dgTask.stream().forEach(t -> assignedPartitions.addAll(t.getPartitionsV2()));
+      totalTaskCount += dgTask.size();
+    }
+
+    List<String> unassignedPartitions = new ArrayList<>(datastreamPartitions.getPartitions());
+    unassignedPartitions.removeAll(assignedPartitions);
+
+    int maxPartitionPerTask = datastreamPartitions.getPartitions().size() / totalTaskCount;
+    // calculate how many tasks are allowed to have slightly more partitions
+    final AtomicInteger remainder = new AtomicInteger(datastreamPartitions.getPartitions().size() % totalTaskCount);
```

Comment:
looks like remainder is used to calculate whether maxPartitionsPerTask should be incremented. Can we use ceil to get maxPartitionsPerTask ? It will make it slightly easier to understand.

Reply:
Assume we have total N tasks, the maxPartitionsPerTask is k. R is the remainder.

We will have R tasks with k partitions and (N-R) task with k -1 partitions.

I haved struct in a way a task will be iterate once so I need to know if this task belongs to R(with k partitions) or (N-R) with (k-1) partitions. 

A ceil to get k will probably not work here since I need to know the task belongs to R or (N-R)